internals/overlord/logstate/gatherer.go

```diff
@@ -0,0 +1,348 @@
+		pullers:    newPullerGroup(target.Name),
+		timer:      newTimer(),
+	}
+	g.clientCtx, g.clientCancel = context.WithCancel(context.Background())
+	g.tomb.Go(g.loop)
+	g.tomb.Go(g.pullers.tomb.Wait)
+
+	return g, nil
+}
+
+func fillDefaultArgs(args logGathererArgs) logGathererArgs {
+	if args.bufferTimeout == 0 {
+		args.bufferTimeout = bufferTimeout
+	}
+	if args.timeoutFinalFlush == 0 {
+		args.timeoutFinalFlush = timeoutFinalFlush
+	}
+	if args.newClient == nil {
+		args.newClient = newLogClient
+	}
+	return args
+}
+
+// PlanChanged is called by the LogManager when the plan is changed, if this
+// gatherer's target exists in the new plan.
+func (g *logGatherer) PlanChanged(pl *plan.Plan, buffers map[string]*servicelog.RingBuffer) {
+	// Remove old pullers
+	for _, svcName := range g.pullers.Services() {
+		svc, svcExists := pl.Services[svcName]
+		if !svcExists {
+			g.pullers.Remove(svcName)
+			continue
+		}
+
+		tgt := pl.LogTargets[g.targetName]
+		if !svc.LogsTo(tgt) {
+			g.pullers.Remove(svcName)
+		}
+	}
+
+	// Add new pullers
+	for _, service := range pl.Services {
+		target := pl.LogTargets[g.targetName]
+		if !service.LogsTo(target) {
+			continue
+		}
+
+		buffer, bufferExists := buffers[service.Name]
+		if !bufferExists {
+			// We don't yet have a reference to the service's ring buffer
+			// Need to wait until ServiceStarted
+			continue
+		}
+
+		g.pullers.Add(service.Name, buffer, g.entryCh)
+	}
+}
+
+// ServiceStarted is called by the LogManager on the start of a service which
+// logs to this gatherer's target.
+func (g *logGatherer) ServiceStarted(service *plan.Service, buffer *servicelog.RingBuffer) {
+	g.pullers.Add(service.Name, buffer, g.entryCh)
+}
+
+// The main control loop for the logGatherer. loop receives logs from the
+// pullers on entryCh, and writes them to the client. It also flushes the
+// client periodically, and exits when the gatherer's tomb is killed.
+func (g *logGatherer) loop() error {
+	defer g.timer.Stop()
+
+mainLoop:
+	for {
+		select {
+		case <-g.tomb.Dying():
+			break mainLoop
+
+		case <-g.timer.Expired():
+			// Mark timer as unset
+			g.timer.Stop()
+			g.handleClientErr(g.client.Flush(g.clientCtx))
+
+		case entry := <-g.entryCh:
+			g.handleClientErr(g.client.Write(g.clientCtx, entry))
+			g.timer.EnsureSet(g.bufferTimeout)
+		}
+	}
+
+	// Final flush to send any remaining logs buffered in the client
+	// We need to create a new context, as the previous one may have been cancelled.
+	ctx, cancel := context.WithTimeout(context.Background(), g.timeoutFinalFlush)
+	defer cancel()
+	g.handleClientErr(g.client.Flush(ctx))
+	return nil
+}
+
+func (g *logGatherer) handleClientErr(err error) {
+	if err == nil {
+		return
+	}
+
+	backoff := &clienterr.Backoff{}
+	if errors.As(err, &backoff) {
+		logger.Noticef("Target %q: %v", g.targetName, err)
+		g.timer.retryAfter = backoff.RetryAfter
+		return
+	}
+
+	timeout := &clienterr.Timeout{}
+	if errors.As(err, &timeout) {
```

Comment:
Yeah having timeout in a form where you can just `Is` will probably look a lot syntactically nicer?